common/src/main/java/io/netty/util/internal/logging/InternalLoggerFactory.java

```diff
@@ -16,24 +16,63 @@
 
 package io.netty.util.internal.logging;
 
+import java.util.concurrent.atomic.AtomicReference;
+
 /**
- * Creates an {@link InternalLogger} or changes the default factory
- * implementation.  This factory allows you to choose what logging framework
- * Netty should use.  The default factory is {@link Slf4JLoggerFactory}.  If SLF4J
- * is not available, {@link Log4JLoggerFactory} is used.  If Log4J is not available,
- * {@link JdkLoggerFactory} is used.  You can change it to your preferred
- * logging framework before other Netty classes are loaded:
- * <pre>
- * {@link InternalLoggerFactory}.setDefaultFactory({@link Log4JLoggerFactory}.INSTANCE);
- * </pre>
- * Please note that the new default factory is effective only for the classes
- * which were loaded after the default factory is changed.  Therefore,
- * {@link #setDefaultFactory(InternalLoggerFactory)} should be called as early
- * as possible and shouldn't be called more than once.
+ * Creates {@link InternalLogger}s. This factory allows you to choose what logging framework Netty should use. The
+ * default factory is {@link Slf4JLoggerFactory}. If SLF4J is not available, {@link Log4JLoggerFactory} is used. If
+ * Log4J is not available, {@link JdkLoggerFactory} is used. You can change it to your preferred logging framework
+ * before other Netty classes are loaded via {@link #setDefaultFactory(InternalLoggerFactory)}. If you want to change
+ * the logger factory, {@link #setDefaultFactory(InternalLoggerFactory)} must be invoked before any other Netty classes
+ * are loaded. Note that {@link #setDefaultFactory(InternalLoggerFactory)}} can not be invoked more than once.
  */
 public abstract class InternalLoggerFactory {
 
-    private static volatile InternalLoggerFactory defaultFactory;
+    private static final InternalLoggerFactoryHolder HOLDER = new InternalLoggerFactoryHolder();
+
+    /**
+     * This class holds a reference to the {@link InternalLoggerFactory}. The raison d'être for this class is primarily
+     * to aid in testing.
+     */
+    static class InternalLoggerFactoryHolder {
+        private final AtomicReference<InternalLoggerFactory> reference;
+
+        InternalLoggerFactoryHolder() {
+            this(null);
+        }
+
+        InternalLoggerFactoryHolder(final InternalLoggerFactory holder) {
+            this.reference = new AtomicReference<InternalLoggerFactory>(holder);
+        }
+
+        InternalLoggerFactory getFactory() {
+            if (reference.get() == null) {
+                reference.compareAndSet(null, newDefaultFactory(InternalLoggerFactory.class.getName()));
+            }
+            return reference.get();
+        }
+
+        void setFactory(final InternalLoggerFactory factory) {
+            if (factory == null) {
+                throw new NullPointerException("factory");
+            }
+            if (reference.get() != null || !reference.compareAndSet(null, factory)) {
```

Comment:
I'm assuming `reference.get() != null` is a performance optimization? If so, can we do with out it? `setFactory` is only allowed to be called 1 time and in the majority case it will actually add overhead.